### PR TITLE
osd/PG: do not blindly roll forward to log.head

### PR DESCRIPTION
**Pull request by liewegas**
If we are told we can roll forward by the primary, we should only roll
forward as far as the primary says we can.

Fixes: http://tracker.ceph.com/issues/24597
Signed-off-by: Sage Weil <sage@redhat.com>

### CODE REVIEW

**Comment by liewegas** [comment]
@jdurgin This is what makes sense to me.. but I'm not super familiar with the EC code.  There was probably a reason those calls where there?

**Comment by jdurgin** [comment]
@liewegas I think your fix is right. When @dzafman and I discussed it in the context of http://tracker.ceph.com/issues/24192 for some reason we thought the third roll_forward call should be removed too, but I can't recall why and it seems correct to leave it in now. Do you remember David?

**Comment by dzafman** [comment]
@jdurgin @liewegas No, I don't remember now.

**Comment by liewegas** [comment]
@jdurgin the third one is the one that actually does the thing, right?
```
  if (roll_forward_to && *roll_forward_to > pg_log.get_can_rollback_to()) {
    pg_log.roll_forward_to(*roll_forward_to, &rollbacker);
    last_rollback_info_trimmed_to_applied = *roll_forward_to;
  }
```

the second one looks ok to me; it's the backfill target case
```
  if (roll_forward_to && entries.rbegin()->soid > info.last_backfill) {
    pg_log.roll_forward(&rollbacker);
  }
```
(if we're beyond last_backfill we must be a backfill target)

**Comment by liewegas** [comment]
Is 24192 a dup of this bug then?

I think http://tracker.ceph.com/issues/23145 is a dup too!

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-07-11_04:59:04-rados-wip-24597-luminous-distro-basic-smithi/ is the luminous run on this.  
- turned up a problem with the standalone test; fixed
- request slow whitelisting noise
- infra deployment noise
- that timer thread crash we recently fixed (not backported yet i guess?)
- a dead run due to network loss (and daemons subseqeuently stopped)
- the envlibrados rocksdb test that is flaky.

Basically looks good to me.  I am still unhappy that users are seeing this failure mode, though, and we haven't seen it in teuthology

**Comment by liewegas** [comment]
Okay, it now has a reproducer!  Fails on original, passes with this fix.

**Comment by liewegas** [comment]
Yeah the extra coverage is good (exposes another bug reliably) and it's easier than opening another PR.  Don't plan to backport it tho?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-07-11_23:47:35-rados-wip-sage3-testing-2018-07-11-1618-distro-basic-smithi/
looks good.
- cls_rgw thing
- ManifestUnset test failure (unrelated)
- eio injection failure (known issue)
